docs/decisions/infra/0010-feature-flags-system-design.md

```diff
@@ -0,0 +1,70 @@
+# Feature flags system design
+
+* Deciders: @aligg @Nava-JoshLong @lorenyu
+* Date: 2023-11-28
+
+## Context
+
+All projects should have some sort of feature flag mechanism for controlling the release and activation of features. This accelerates product development by unblocking developers from being able to deploy continuously while still providing business owners with control over when features are visible to end users. More advanced feature flag systems can also provide the ability to do gradual rollouts to increasing percentages of end users and to do split tests (also known as A/B tests) to evaluate the impact of different feature variations on user behavior and outcomes, which provide greater flexibility on how to reduce the risk of launching features.
+
+## Requirements
```

Comment:
Thinking about it a bit more from the API side, how would these feature flags actually be implemented?

I looked into feature flags for the grants project, and the actual API implementation had a lot of complications (we’ve only implemented something very basic at the moment). Mainly, if you want to be able to update a feature flag while the service is still running, you need some process for fetching the configuration periodically.

How would you use this technology to do that?

Reply:
For an API, it kinda depends on how the feature flag is being called. In NJ, we have the ID on the client being the user's cognito ID, and we can pass that to the API so that user gets a set experience based on the flag's configuration. 

Here's the two paths I'm seeing for implementing in an API

- The evaluate call is either in a function that polls periodically, like you said
  - We have an app that runs in a while loop, and the evaluate call is in there
  - The time where a user might have the older value is then at most the interval of the loop
  - Best for light switch flags, user experience ones would still need to evaluate in case they were cached before
  - More AWS API calls compared to other option
- Or, be in the handler for the endpoint being called
  - The flag's new value would be seen almost immediately after the change
  - Much less AWS API calls since it only pulls as needed
  - Works for both light switch flags and user experience flags

If you just want a light switch for cutting over to a new process, like we have in a separate app in NJ, you'd need to tweak the evaluate call to have the caller ID as a timestamp. That way, every evaluate isn't cached on AWS' side, and it will always pull the new value

Reply:
I made a very basic version that loads from env var on app initialization + lets you override with headers (similar to some of what we did on NJ UI) on the grants project (open source, so should be viewable): https://github.com/HHS/simpler-grants-gov/commit/e5f36255b0ab66579497ff1fd93cc95a9e3e9b57

I set it up in a way that we could plug in a background thread that would pull from _somewhere_ and update that configuration live if desired. Apparently adding that kind of background thread with our already multi-threaded WSGI app causes some issues (gotta configure uvicorn to load the app first and then split), but its doable. As we don't yet have any use case for feature flags, I haven't gone any further with the implementation

Reply:
I'm having a hard time following this thread.

This is how the feature flags will be used:

```mermaid
sequenceDiagram
    Browser->>Server: HTTP request
    Server->>Evidently: EvaluateFeature(feature, userId)
    Evidently->>Server: FeatureOn | FeatureOff
    Server->>Browser: Render appropriate experience
```

if a business owner turns on/off a feature, it will be reflected in a subsequent HTTP request

Reply:
Lets say I've built an API in our template repo that has a feature flag, how will that work?

I'm not following if what you're referring to in that diagram as a server is our API server, or something else. How does the API know the value has changed live (if possible)? Will we need to add a call in every API request to get the current values?

Or, put another way, what changes would follow-up this infra change in our API repo?

Reply:
I'm not sure I follow completely the challenges we're talking about here. 

We can integrate w/ Evidently and make a call for a feature variation's value. This means we can change the feature flag at any time without needing a deploy. 

Why would we need polling?

Reply:
> I'm not following if what you're referring to in that diagram as a server is our API server, or something else.

Yeah the server would be the API server if it's an API, or Next.js server if we're talking about the Next.js user-facing application.

> How does the API know the value has changed live (if possible)? Will we need to add a call in every API request to get the current values?

Yeah, we'd call Evidently on each API request that goes through code path that involves that feature flag.

Reply:
> what changes would follow-up this infra change in our API repo?

i sketched it out in this spec, but we might implement a `feature_flag_manager` module that looks something like this (pseudocode):

```python
import boto3

class FeatureFlagManager:
  client: EvidentlyClient

  def __init__(self):
    self.client = boto3.client('evidently')

  def is_feature_enabled(feature: str, user_id: str) -> bool:
    response = self.client.evaluate_feature(feature, entity_id=user_id)
    return response.value == "FeatureOn"
```